### DISCUSSION TITLE
ML-based LS Operator Selection

### DISCUSSION THREAD

**Discussion by inboxedshoe** [google/or-tools#2566] [Routing (and legacy CP) questions]
I've been looking into the GoogleOR documentation and trying to find a way to change or keep the local search operator being used at any given step. I looked into search monitors and then the metaheuristic models but couldn't seem to find a way to select the operator to use mid-way. I'm not certain if this route is feasible as the examples don't show a method of switching LS Operators either.

The main concern was to also be able to use the routing model. I looked into the simplest method of restarting the search or setting up a new one every time with a 1-step limit but that caused unneeded overhead and was often buggy, especially when accessing the routing model through the python wrapper. 

Any thoughts on other steps I can try to customize the model? I'm obviously quite new to GoogleOR as a whole and hoping here would be a good place to get a proper foothold.

**Comment by Mizux** [comment]
Do you mean changing this parameter during the search ?
https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/constraint_solver/samples/vrp_with_time_limit.py#L99-L100

What about running in parallel the same model with different LS metaheuristic ?

**Reply by inboxedshoe** [reply]
There's that But I also want to be able to change the local search (ex. do 1 step 2-opt, check result, continue 2-opt or change to 3-opt etc). This technically would make a parallel approach invalid.

Which is basically in this part of the search parameters:
<code><pre>
local_search_operators {
  use_relocate: true
  use_relocate_pair: true
  use_exchange: true
  use_cross: true
  use_two_opt: true
  use_or_opt: true
  use_lin_kernighan: true
  use_make_active: true
  use_make_inactive: true
  use_swap_active: true
  use_path_lns: true
  use_inactive_lns: true
}</code></pre>

**Comment by inboxedshoe** [comment]
As an update, I tried accessing the routing model's underlying solver, adding a decision builder, and calling newSearch but to no response from the code. I'm guessing that function isn't available to call.
